stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb

```diff
@@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
     return base._copyToNativeArrayBuffer()
   }
 
-  /// Copy a Sequence into an array.
-  public func _initializeTo(ptr: UnsafeMutablePointer<Base.Generator.Element>)
-    -> UnsafeMutablePointer<Base.Generator.Element> {
+  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
+  /// returning one past the last element initialized, or a `nil` pointer
+  /// if the copy cannot be safely made given `capacity`.
+  public func _initializeTo(
```

Comment:
on Mon Mar 07 2016, Károly Lőrentey <notifications-AT-github.com> wrote:

>> @@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
>>      return base._copyToNativeArrayBuffer()
>>    }
>>  
>> -  /// Copy a Sequence into an array.
>> -  public func _initializeTo(ptr: UnsafeMutablePointer<Base.Generator.Element>)
>> -    -> UnsafeMutablePointer<Base.Generator.Element> {
>> +  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
>> +  /// returning one past the last element initialized, or a `nil` pointer
>> +  /// if the copy cannot be safely made given `capacity`.
>> +  public func _initializeTo(
>
> `capacity` is the number of uninitialized elements that the caller
> made available in the buffer that `ptr` points to; indeed, this should
> be spelled out in the doc comment.

It doesn't read well at the call site because it isn't clear what it's
the capacity of.  Anywhere you are going to pass an
`UnsafeMutablePointer` + length of its potential storage, it might be
better to use an `UnsafeMutableBufferPointer`... although that was not
really designed to be used with uninitialized memory, you can always get
the baseAddress.

> The parameter is necessary because `_initializeTo` is defined on
> `SequenceType`, where there is no guaranteed `count` — so the caller
> might not be able to allocate enough space to fit all elements. An
> alternative would be to trust that the caller allocates enough space
> for `underestimatedCount` elements, but that didn't feel like the
> right API to me. Having the capacity explicitly supplied is harder to
> get wrong, and (in theory) allows the function to work in cases when
> the underestimate was low but the caller happened to allocate more
> space anyway due to e.g. fitting storage size on an exponential growth
> curve.

This makes sense to me... but only if we really need this method on
`SequenceType`.  Do we?  Otherwise it's premature generalization.

> The `_initializeTo` in this PR is somewhat limited by only allowing
> the sequence to be copied into a single contiguous buffer. This pretty
> much limits its use to `Array`. I'd actually prefer to see a
> non-failing but continuable variant that would allow its use in any
> collection that segments its storage into multiple contiguous regions
> (e.g., a deque based on a circular buffer, or even my B-tree
> collections). This would make the method signature even more
> complicated in `Sequence`, though. OTOH, it would fit a bit better
> into `IteratorProtocol`, with a declaration that could be a variant of
> this:
>
> ```
> /// Copy at most `capacity` elements starting from the current element into
> /// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
> /// - Returns: `true` if there are more elements remaining in the sequence.
> mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
> ```

Why do you think this belongs in `IteratorProtocol`?

> (I don't like the `inout` parameter or the boolean result though.)

Me neither.  Why do it that way?

> However, having `_copyContents` in `IteratorProtocol` would probably
> mean we'd need to change `Array` to use a specialized iterator other
> than `IndexingIterator`. 

...because?

> This would mean writing up a public API proposal that'd be hard to
> justify unless `copyContents` was made public at the same time — and
> I'm not yet sure if the speedup would be significant enough to be
> worth that effort.
>
> I'd  be OK with any of the following options:
>
> 1. Leave the `capacity` parameter in place (possibly with a better label), but explicitly document what it does.
> 2. Drop the parameter, but document that the caller is supposed to
> supply enough space to fit `underestimatedCount` elements. The method
> would still need to be able to fail, though.
> 3. Move the method down to `Collection`, restoring it to the original
> signature, but making it less useful.

Presumably you'd need to do something else for `SequenceType` somewhere?
Otherwise, we'd be introducing complexity to the signature just for
theoretical utility with `SequenceType` that we're not actually taking
advantage of, which sounds like a bad idea.

> 4. Drop the `_initializeTo` part of this PR, and concentrate on
> replacing indexing with iterators (along with speeding up iterators
> when necessary).  5. Do more work to convert `_initializeTo` into a
> restartable/continuable thing, possibly moving it into
> `IteratorProtocol`.
>
> Thoughts?

IIUC the main point of this PR is that using iterators is better than
indexing; if that is separable from the `_initializeTo` change, I'd much
rather accept that part alone, and consider other changes in a different
PR.

-- 
-Dave